torch/csrc/autograd/c2isl_function.cpp

```diff
@@ -0,0 +1,321 @@
+// anything about const-ness.
+int64_t* newDLInt64Array(const at::IntList& arr) {
+  auto len = arr.size();
+  auto r = new int64_t[len];
+  for (size_t i = 0; i < len; i++) {
+    r[i] = arr[i];
+  }
+  return r;
+}
+
+DLMetadataUPtr toDLMetadata(at::Tensor& t) { // should be const
+  DLMetadataUPtr res(new DLMetadata);
+  res->ndim = t.dim();
+  res->dtype = toDLDataType(t.type());
+  res->shape = newDLInt64Array(t.sizes()); // ownership transfer
+  res->strides = newDLInt64Array(t.strides()); // ownership transfer
+  // invalid values which should not be used
+  res->data = nullptr;
+  res->ctx.device_type = DLDeviceType(0);
+  res->ctx.device_id = -1;
+  res->byte_offset = 0;
+  return res;
+}
+
+// NB: DLTensorUPtr is only live as long as Tensor is
+// TODO: check ownership against numpy (in cwrap)
+DLTensorUPtr toDLTensor(at::Tensor& t) {
+  DLTensorUPtr res(new DLTensor, DLTensorDeleter(t));
+  // NB: It is NOT sound to resize afterwards.  It would be good to clear
+  // TH_STORAGE_RESIZABLE but ATen does not currently expose this
+  // functionality, see https://github.com/zdevito/ATen/issues/28
+  res->data = t.data_ptr();
+  if (t.type().isCuda()) {
+    res->ctx.device_type = DLDeviceType::kGPU;
+    res->ctx.device_id = t.get_device();
+  } else {
+    res->ctx.device_type = DLDeviceType::kCPU;
+    // TODO: It's not clear what the semantics of device_id are supposed
+    // to be on CPU
+    res->ctx.device_id = 0;
+  }
+  res->ndim = t.dim();
+  res->dtype = toDLDataType(t.type());
+  // It would be nice to directly use the data() of these
+  // objects, but ATen is const and dlpack is non-const, which
+  // means a client could scribble over the data. Safer to not.
+  res->shape = newDLInt64Array(t.sizes()); // ownership transfer
+  res->strides = newDLInt64Array(t.strides()); // ownership transfer
+  // NB: byte_offset is already applied by THTensor_(data), which is what backs
+  // data_ptr.
+  res->byte_offset = 0;
+  return res;
+}
+
+// TODO: Replace this with the direct thing when ATen replaces thpp
+at::Tensor variableToATen(const std::shared_ptr<Variable>& var) {
+  THPObjectPtr pyvar(THPVariable_Wrap(var));
+  PyObject* data = ((THPVariable*)pyvar.get())->data;
+  return createTensorAT(data);
+}
+
+variable_list ISLFunction::apply(const variable_list& input_vars) {
+  AutoGIL gil;
+
+  // Use the same backend as the first tensor
+  auto& input = input_vars.at(0);
+  AutoGPU guard(input->data->getDevice());
+
+  if (!pImpl_) {
+    std::vector<const DLMetadata*> inMetas; // ownership managed by inTensorUPtrs
+    std::vector<DLMetadataUPtr> inMetaUPtrs;
+    for (auto input : input_vars) {
+      auto at = variableToATen(input);
+      inMetaUPtrs.emplace_back(toDLMetadata(at));
+      inMetas.emplace_back(inMetaUPtrs.back().get());
+    }
+    initTVM(inMetas);
+    pImpl_ = std::unique_ptr<c2isl::ISLTVMIROp>(
+              new c2isl::ISLTVMIROp(outputs, inputs, vars, ops)
```

Comment:
I gave them these names so we lined up with c2isl code better, but now that we have a Python interface this is less pressing, so I'll do this.